librad/src/git/trailer.rs

```diff
@@ -0,0 +1,139 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::convert::TryFrom;
+
+#[derive(Debug, Clone, Eq, PartialEq, thiserror::Error)]
+pub enum Error {
+    #[error("the given input is not a well-formed trailer")]
+    IllFormed,
+
+    #[error("could not parse the trailer due to an invalid token found")]
+    Token(#[from] InvalidToken),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Trailer {
+    token: Token,
+    value: String,
```

Comment:
Since we will either not touch the trailer at all, or parse the value into some other representation, it would be good to leave `value` at `&str`.

Reply:
That's a good point.

Reply:
Wait, I actually had decided to use `String` since using `&str` requires the whole lifetime parameter bloat, which cascades. Is your concern purely on performance?

Reply:
And what's the problem with lifetime parameters? Copying lots of small strings is almost always bad, and we'll be doing this a lot.

I was thinking that keeping `Token` owned would be necessary because of case-folding, but it looks like [unicase](https://docs.rs/unicase/2.6.0/unicase/) works fine with borrowed values, so we can keep it `&str`, too.

Reply:
Oh well, disregard my comments about `FromStr` then -- you'll indeed need `TryFrom` when the types have lifetimes.